File: 290-Word-Pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        pattern, s = list(pattern), s.split(' ')

        if len(pattern) != len(s):
            return False

        hashPS, hashSP = {}, {}

        for i in range(len(s)):
            if pattern[i] in hashPS and hashPS[pattern[i]] != s[i]:
                return False
            hashPS[pattern[i]] = s[i]

            if s[i] in hashSP and hashSP[s[i]] != pattern[i]:
                return False
            hashSP[s[i]] = pattern[i]

        return True
        

File: neetcode/290-Word-Pattern.ts
function wordPattern(pattern: string, s: string): boolean {
  const words = s.split(' ');
  if (pattern.length !== words.length) return false;

  const charToWord = new Map<string, string>();
  const wordToChar = new Map<string, string>();

  for (let i = 0; i < pattern.length; i++) {
    const char = pattern[i];
    const word = words[i];

    if (charToWord.has(char) && charToWord.get(char) !== word) return false;
    charToWord.set(char, word);

    if (wordToChar.has(word) && wordToChar.get(word) !== char) return false;
    wordToChar.set(word, char);
  }

  return true;
}